File: backend/services/reading_progress.py
from datetime import datetime, date
from typing import List, Optional, Dict
from sqlmodel import Session, select
from sqlalchemy import func
import json

try:
    from backend.models import Edition, ReadingProgress, ReadingStats, Book
    from backend.database import get_db_session
except ImportError:
    from models import Edition, ReadingProgress, ReadingStats, Book
    from database import get_db_session


class ReadingProgressService:
    def __init__(self):
        pass

    def update_progress(self, edition_id: int, current_page: int, total_pages: int) -> ReadingProgress:
        """Update reading progress for a book"""
        with get_db_session() as session:
            progress = session.exec(
                select(ReadingProgress).where(
                    ReadingProgress.edition_id == edition_id,
                    ReadingProgress.user_id == 1
                )
            ).first()
            
            if not progress:
                progress = ReadingProgress(
                    user_id=1,
                    edition_id=edition_id,
                    status="currently_reading"
                )
                session.add(progress)
            
            progress.current_page = current_page
            progress.total_pages = total_pages
            progress.progress_percentage = (current_page / total_pages * 100) if total_pages > 0 else 0
            progress.updated_at = datetime.utcnow()
            
            # If book wasn't being read, set start date
            if progress.status != "currently_reading":
                progress.status = "currently_reading"
                progress.start_date = datetime.utcnow()
            
            session.commit()
            session.refresh(progress)
            return progress

    def get_reading_stats(self) -> ReadingStats:
        """Get reading statistics for the user"""
        with get_db_session() as session:
            stats = ReadingStats()

            # Count books by status
            stats.books_read = session.exec(
                select(func.count(ReadingProgress.id)).where(
                    ReadingProgress.user_id == 1,
                    ReadingProgress.status == "finished"
                )
            ).first() or 0

            stats.books_reading = session.exec(
                select(func.count(ReadingProgress.id)).where(
                    ReadingProgress.user_id == 1,
                    ReadingProgress.status == "currently_reading"
                )
            ).first() or 0

            stats.books_want_to_read = session.exec(
                select(func.count(ReadingProgress.id)).where(
                    ReadingProgress.user_id == 1,
                    ReadingProgress.status == "want_to_read"
                )
            ).first() or 0

            # Calculate total pages read
            finished_books = session.exec(
                select(ReadingProgress).where(
                    ReadingProgress.user_id == 1,
                    ReadingProgress.status == "finished",
                    ReadingProgress.total_pages.isnot(None)
                )
            ).all()

            stats.total_pages_read = sum(book.total_pages for book in finished_books if book.total_pages)

            # Calculate average rating
            ratings = session.exec(
                select(ReadingProgress.rating).where(
                    ReadingProgress.user_id == 1,
                    ReadingProgress.rating.isnot(None)
                )
            ).all()

            if ratings:
                stats.average_rating = sum(ratings) / len(ratings)

            # Books read this month and year
            current_date = date.today()
            month_start = date(current_date.year, current_date.month, 1)
            year_start = date(current_date.year, 1, 1)

            stats.books_read_this_month = session.exec(
                select(func.count(ReadingProgress.id)).where(
                    ReadingProgress.user_id == 1,
                    ReadingProgress.status == "finished",
                    ReadingProgress.finish_date >= month_start
                )
            ).first() or 0

            stats.books_read_this_year = session.exec(
                select(func.count(ReadingProgress.id)).where(
                    ReadingProgress.user_id == 1,
                    ReadingProgress.status == "finished",
                    ReadingProgress.finish_date >= year_start
                )
            ).first() or 0

            # Calculate genre breakdown
            stats.genre_breakdown, stats.genre_percentages = self._calculate_genre_breakdown(session, finished_books)

            return stats

    def get_books_by_status(self, status: str) -> List[ReadingProgress]:
        """Get all books with a specific reading status"""
        with get_db_session() as session:
            return session.exec(
                select(ReadingProgress).where(
                    ReadingProgress.user_id == 1,
                    ReadingProgress.status == status
                )
            ).all()

    def start_reading(self, edition_id: int) -> ReadingProgress:
        """Mark a book as currently reading"""
        with get_db_session() as session:
            progress = session.exec(
                select(ReadingProgress).where(
                    ReadingProgress.edition_id == edition_id,
                    ReadingProgress.user_id == 1
                )
            ).first()
            
            if not progress:
                progress = ReadingProgress(
                    user_id=1,
                    edition_id=edition_id
                )
                session.add(progress)
            
            progress.status = "currently_reading"
            progress.start_date = datetime.utcnow()
            progress.updated_at = datetime.utcnow()
            
            session.commit()
            session.refresh(progress)
            return progress

    def finish_reading(self, edition_id: int, rating: Optional[int] = None) -> ReadingProgress:
        """Mark a book as finished"""
        with get_db_session() as session:
            progress = session.exec(
                select(ReadingProgress).where(
                    ReadingProgress.edition_id == edition_id,
                    ReadingProgress.user_id == 1
                )
            ).first()
            
            if not progress:
                progress = ReadingProgress(
                    user_id=1,
                    edition_id=edition_id
                )
                session.add(progress)
            
            progress.status = "finished"
            progress.finish_date = datetime.utcnow()
            progress.updated_at = datetime.utcnow()
            
            if rating:
                progress.rating = rating
            
            # Set progress to 100% if pages are known
            if progress.total_pages:
                progress.current_page = progress.total_pages
                progress.progress_percentage = 100.0
            
            session.commit()
            session.refresh(progress)
            return progress

    def add_to_wishlist(self, edition_id: int) -> ReadingProgress:
        """Add a book to the want-to-read list"""
        with get_db_session() as session:
            progress = session.exec(
                select(ReadingProgress).where(
                    ReadingProgress.edition_id == edition_id,
                    ReadingProgress.user_id == 1
                )
            ).first()

            if not progress:
                progress = ReadingProgress(
                    user_id=1,
                    edition_id=edition_id,
                    status="want_to_read"
                )
                session.add(progress)
            else:
                progress.status = "want_to_read"
                progress.updated_at = datetime.utcnow()

            session.commit()
            session.refresh(progress)
            return progress

    def _calculate_genre_breakdown(self, session: Session, finished_books: List[ReadingProgress]) -> tuple:
        """Calculate genre breakdown from finished books"""
        genre_count: Dict[str, int] = {}

        for progress in finished_books:
            edition = session.exec(
                select(Edition).where(Edition.id == progress.edition_id)
            ).first()

            if not edition:
                continue

            book = session.exec(
                select(Book).where(Book.id == edition.book_id)
            ).first()

            if not book or not book.genres:
                continue

            try:
                genres = json.loads(book.genres) if isinstance(book.genres, str) else book.genres
                if isinstance(genres, list):
                    for genre in genres:
                        genre_str = str(genre).strip()
                        if genre_str:
                            genre_count[genre_str] = genre_count.get(genre_str, 0) + 1
            except (json.JSONDecodeError, TypeError):
                pass

        # Calculate percentages
        genre_percentages: Dict[str, float] = {}
        total_books = len(finished_books)

        if total_books > 0:
            for genre, count in genre_count.items():
                percentage = (count / total_books) * 100
                genre_percentages[genre] = round(percentage, 2)

        return genre_count, genre_percentages

File: v2/apps/web/src/lib/services/reading-progress.ts
import { db } from '../db';
import { readingProgress, readingGoals, books, editions, authors, bookAuthors, userBooks } from '@booktarr/database';
import { eq, and, desc, sql, gte, lte } from 'drizzle-orm';

export type ReadingStatus = 'want_to_read' | 'currently_reading' | 'finished' | 'dnf' | 'on_hold';

export interface CreateReadingProgressInput {
  userId: string;
  bookId: string;
  status?: ReadingStatus;
  currentPage?: number;
  totalPages?: number;
  rating?: number;
  review?: string;
}

export interface UpdateReadingProgressInput {
  status?: ReadingStatus;
  currentPage?: number;
  totalPages?: number;
  progressPercentage?: number;
  rating?: number;
  review?: string;
  reviewPublic?: 'private' | 'friends' | 'public';
}

export interface ReadingStats {
  totalBooksRead: number;
  currentlyReading: number;
  wantToRead: number;
  booksFinishedThisYear: number;
  booksFinishedThisMonth: number;
  averageRating: number;
  totalPagesRead: number;
  totalReadingTime: number;
  favoriteGenres: string[];
}

export class ReadingProgressService {
  /**
   * Get or create reading progress for a book
   */
  async getOrCreateProgress(userId: string, bookId: string): Promise<any> {
    const existing = await db
      .select()
      .from(readingProgress)
      .where(and(eq(readingProgress.userId, userId), eq(readingProgress.bookId, bookId)))
      .limit(1);

    if (existing.length > 0) {
      return existing[0];
    }

    const [progress] = await db
      .insert(readingProgress)
      .values({
        userId,
        bookId,
        status: 'want_to_read',
        currentPage: 0,
        progressPercentage: 0,
        readingSessions: 0,
        totalReadingTime: 0,
      })
      .returning();

    return progress;
  }

  /**
   * Update reading progress
   */
  async updateProgress(
    userId: string,
    bookId: string,
    input: UpdateReadingProgressInput
  ): Promise<any> {
    // Calculate progress percentage if pages are provided
    let progressPercentage = input.progressPercentage;
    if (input.currentPage !== undefined && input.totalPages !== undefined && input.totalPages > 0) {
      progressPercentage = Math.round((input.currentPage / input.totalPages) * 100);
    }

    // If status is changing to currently_reading and no startedAt, set it
    const updates: any = {
      ...input,
      updatedAt: new Date(),
      lastReadAt: new Date(),
    };

    if (progressPercentage !== undefined) {
      updates.progressPercentage = progressPercentage;
    }

    // Get existing progress to check status changes
    const existing = await this.getOrCreateProgress(userId, bookId);

    if (input.status === 'currently_reading' && !existing.startedAt) {
      updates.startedAt = new Date();
    }

    if (input.status === 'finished' && !existing.finishedAt) {
      updates.finishedAt = new Date();
      updates.progressPercentage = 100;
      updates.currentPage = input.totalPages || existing.totalPages;
    }

    const [updated] = await db
      .update(readingProgress)
      .set(updates)
      .where(and(eq(readingProgress.userId, userId), eq(readingProgress.bookId, bookId)))
      .returning();

    return updated;
  }

  /**
   * Start reading a book
   */
  async startReading(userId: string, bookId: string, totalPages?: number): Promise<any> {
    const progress = await this.getOrCreateProgress(userId, bookId);

    return this.updateProgress(userId, bookId, {
      status: 'currently_reading',
      totalPages: totalPages || progress.totalPages || undefined,
      currentPage: progress.currentPage || 0,
    });
  }

  /**
   * Finish reading a book
   */
  async finishReading(
    userId: string,
    bookId: string,
    rating?: number,
    review?: string
  ): Promise<any> {
    const progress = await this.getOrCreateProgress(userId, bookId);

    return this.updateProgress(userId, bookId, {
      status: 'finished',
      currentPage: progress.totalPages || progress.currentPage,
      rating,
      review,
    });
  }

  /**
   * Get currently reading books for a user
   */
  async getCurrentlyReading(userId: string): Promise<any[]> {
    const results = await db
      .select({
        readingProgress: readingProgress,
        book: books,
        edition: editions,
        userBook: userBooks,
      })
      .from(readingProgress)
      .innerJoin(books, eq(readingProgress.bookId, books.id))
      .innerJoin(editions, eq(books.id, editions.bookId))
      .innerJoin(userBooks, and(
        eq(userBooks.editionId, editions.id),
        eq(userBooks.userId, userId)
      ))
      .where(and(
        eq(readingProgress.userId, userId),
        eq(readingProgress.status, 'currently_reading')
      ))
      .orderBy(desc(readingProgress.lastReadAt));

    // Get authors for each book
    const booksWithAuthors = await Promise.all(
      results.map(async (result) => {
        const bookAuthorsData = await db
          .select({
            author: authors,
            role: bookAuthors.role,
          })
          .from(bookAuthors)
          .innerJoin(authors, eq(bookAuthors.authorId, authors.id))
          .where(eq(bookAuthors.bookId, result.book.id))
          .orderBy(bookAuthors.displayOrder);

        return {
          userBook: result.userBook,
          edition: result.edition,
          book: result.book,
          authors: bookAuthorsData.map((ba) => ba.author),
          readingProgress: result.readingProgress,
        };
      })
    );

    return booksWithAuthors;
  }

  /**
   * Get books by status
   */
  async getBooksByStatus(userId: string, status: ReadingStatus): Promise<any[]> {
    const results = await db
      .select({
        progress: readingProgress,
        book: books,
      })
      .from(readingProgress)
      .innerJoin(books, eq(readingProgress.bookId, books.id))
      .where(and(eq(readingProgress.userId, userId), eq(readingProgress.status, status)))
      .orderBy(desc(readingProgress.updatedAt));

    return results;
  }

  /**
   * Get reading statistics for a user
   */
  async getReadingStats(userId: string): Promise<ReadingStats> {
    const currentYear = new Date().getFullYear();
    const currentMonth = new Date().getMonth();

    // Get all reading progress for user
    const allProgress = await db
      .select()
      .from(readingProgress)
      .where(eq(readingProgress.userId, userId));

    // Calculate stats
    const totalBooksRead = allProgress.filter((p) => p.status === 'finished').length;
    const currentlyReading = allProgress.filter((p) => p.status === 'currently_reading').length;
    const wantToRead = allProgress.filter((p) => p.status === 'want_to_read').length;

    const booksFinishedThisYear = allProgress.filter(
      (p) =>
        p.status === 'finished' &&
        p.finishedAt &&
        new Date(p.finishedAt).getFullYear() === currentYear
    ).length;

    const booksFinishedThisMonth = allProgress.filter(
      (p) =>
        p.status === 'finished' &&
        p.finishedAt &&
        new Date(p.finishedAt).getFullYear() === currentYear &&
        new Date(p.finishedAt).getMonth() === currentMonth
    ).length;

    const ratingsSum = allProgress.reduce((sum, p) => sum + (p.rating || 0), 0);
    const ratingsCount = allProgress.filter((p) => p.rating !== null).length;
    const averageRating = ratingsCount > 0 ? ratingsSum / ratingsCount : 0;

    const totalPagesRead = allProgress.reduce((sum, p) => {
      if (p.status === 'finished' && p.totalPages) {
        return sum + p.totalPages;
      }
      if (p.status === 'currently_reading' && p.currentPage) {
        return sum + p.currentPage;
      }
      return sum;
    }, 0);

    const totalReadingTime = allProgress.reduce((sum, p) => sum + (p.totalReadingTime || 0), 0);

    return {
      totalBooksRead,
      currentlyReading,
      wantToRead,
      booksFinishedThisYear,
      booksFinishedThisMonth,
      averageRating: Math.round(averageRating * 10) / 10,
      totalPagesRead,
      totalReadingTime,
      favoriteGenres: [], // TODO: Implement genre tracking
    };
  }

  /**
   * Get reading progress for a specific book
   */
  async getProgressForBook(userId: string, bookId: string): Promise<any | null> {
    const results = await db
      .select()
      .from(readingProgress)
      .where(and(eq(readingProgress.userId, userId), eq(readingProgress.bookId, bookId)))
      .limit(1);

    return results.length > 0 ? results[0] : null;
  }

  /**
   * Delete reading progress
   */
  async deleteProgress(userId: string, bookId: string): Promise<void> {
    await db
      .delete(readingProgress)
      .where(and(eq(readingProgress.userId, userId), eq(readingProgress.bookId, bookId)));
  }

  /**
   * Get or create reading goal for a year
   */
  async getOrCreateGoal(userId: string, year: number): Promise<any> {
    const existing = await db
      .select()
      .from(readingGoals)
      .where(and(eq(readingGoals.userId, userId), eq(readingGoals.year, year)))
      .limit(1);

    if (existing.length > 0) {
      return existing[0];
    }

    const [goal] = await db
      .insert(readingGoals)
      .values({
        userId,
        year,
        targetBooks: 50, // Default goal
        booksRead: 0,
        pagesRead: 0,
      })
      .returning();

    return goal;
  }

  /**
   * Update reading goal
   */
  async updateGoal(userId: string, year: number, targetBooks: number, targetPages?: number): Promise<any> {
    await this.getOrCreateGoal(userId, year);

    const [updated] = await db
      .update(readingGoals)
      .set({
        targetBooks,
        targetPages,
        updatedAt: new Date(),
      })
      .where(and(eq(readingGoals.userId, userId), eq(readingGoals.year, year)))
      .returning();

    return updated;
  }
}
